### PR TITLE
Update MathTests.swift

### PR DESCRIPTION
**Pull request by anigasan**
Adding test function for cosineDistance

### CODE REVIEW

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 1]
```diff
@@ -199,6 +199,13 @@ final class MathOperatorTests: XCTestCase {
         let output: Float = 1.0
         XCTAssertEqual(z, Tensor(output))
     }
+    func testCosineDistance(){
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Can you add an empty line before the function? Also 
```suggestion
    func testCosineDistance() {
```

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 2]
```diff
@@ -514,6 +521,7 @@ final class MathOperatorTests: XCTestCase {
         ("testIsFinite", testIsFinite),
         ("testIsInfinite", testIsInfinite),
         ("testIsNaN", testIsNaN),
+        ("testCosineDistance", testCosineDistance),
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Thanks @anigasan! Since you defined this under `testCosineSimilarity` above, please also put this line under the `testCosineSimilarity` line here.

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 3]
```diff
@@ -58,6 +58,84 @@ final class MathOperatorTests: XCTestCase {
...
+    func testSquaredDifference() {
+        let x = Tensor<Float>([-5.8])
+        let y = Tensor<Float>([-5.7])
+        let z = squaredDifference(x, y)
+        let zaprox = 132.25
+        assertEqual(z, zaprox, accuracy: 0.0001)
+    }
+    
+    func testZeros() {
+        let x = Tensor<Float>(zeros:[1])
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]:
```suggestion
        let x = Tensor<Float>(zeros: [1])
```

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 4]
```diff
@@ -58,6 +58,84 @@ final class MathOperatorTests: XCTestCase {
...
+    }
+    
+    func testLogSoftmax() {
+        let x = Tensor<Float>([32.0, 34.0, 35.0])
+        let y = logSoftmax(x)
+        let y1 = Tensor<Float>([-3.34901, -1.34901, -0.34901])
+        assertEqual(y, y1, accuracy: 0.0001)
+    }
+    
+    func testLogCoshLoss() {
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: We already test `logCoshLoss` in `LossTests.swift`, so i think this can be removed.

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 5]
```diff
@@ -58,6 +58,84 @@ final class MathOperatorTests: XCTestCase {
...
+        let y = Tensor<Float>([57.0])
+        let z = min(x, y)
+        let zExpected = 57.0
+        assertEqual(z, zExpected)
+    }
+    
+    func testRound() {
+        let x = Tensor<Float>([58.76])
+        let y = round(x)
+        let yExp = 58.0
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Are you sure the value isn't rounded to 59 instead? That would be the nearest integer to 58.76

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 6]
```diff
@@ -58,6 +58,84 @@ final class MathOperatorTests: XCTestCase {
...
+        let y = logSoftmax(x)
+        let y1 = Tensor<Float>([-3.34901, -1.34901, -0.34901])
+        assertEqual(y, y1, accuracy: 0.0001)
+    }
+    
+    func testLogCoshLoss() {
+        let predicted = Tensor<Float>([32.0, 34.0, 36.0])
+        let expected = Tensor<Float>([32.0, 36.0, 37.0])
+        let val = logCoshLoss(predicted:predicted , expected:expected)
+        let valExpected = 0.58626
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Also usually, in tests while writing the expected values, expected precedes the variable name. So in this case it would be, `expectedVal` instead of `valExpected`.

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 7]
```diff
@@ -58,6 +58,84 @@ final class MathOperatorTests: XCTestCase {
...
+        assertEqual(val, valExpected, accuracy: 0.0001)
+    }
+    
+    func testMax() {
+        let x = Tensor<Float>([58.0])
+        let y = Tensor<Float>([57.0])
+        let z = max(x, y)
+        let zExpected = 58.0
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]:
```suggestion
        let expectedZ= 58.0
```
And the subsequent assertion should be updated as well. Same for all other tests.

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 8]
```diff
@@ -58,6 +58,76 @@ final class MathOperatorTests: XCTestCase {
...
+        let x = Tensor<Float>([-1.0])
+        let y = abs(x)
+        assertEqual(y, Tensor<Float>([1.0]))
+    }
+    
+    func testSquaredDifference() {
+        let x = Tensor<Float>([-5.8])
+        let y = Tensor<Float>([-5.7])
+        let z = squaredDifference(x, y)
+        let aproxZ = 132.25
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Nit: I think this would be `approxZ`.

**Comment by Shashi456** [comment]
Also, thank you for adding these tests :).

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift (hunk 8 above)
Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Sorry about that! will make the changes promptly

**Comment by anigasan** [comment]
It is no problem, happy to contribute :)

**Comment by Shashi456** [comment]
Umm @dan-zheng I can't put a build on this. But this looks ready to go.

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 9]
```diff
@@ -58,6 +58,76 @@ final class MathOperatorTests: XCTestCase {
...
+        let y = abs(x)
+        assertEqual(y, Tensor<Float>([1.0]))
+    }
+    
+    func testSquaredDifference() {
+        let x = Tensor<Float>([-5.8])
+        let y = Tensor<Float>([-5.7])
+        let z = squaredDifference(x, y)
+        let approxZ = 132.25
+        assertEqual(z, aproxZ, accuracy: 0.0001)
```
Comment (Shashi456) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: @anigasan you forgot changing the variable name here.  Could you run `swift build` and `swift test` locally and tell if it passes or not?

Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: I'll check it out and see if it passes

Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: I made some small arithmetic mistakes which may have affected the build. I fixed them now

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift (hunk 2 above)
Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: I have made the necessary changes needed

**Comment by Shashi456** [comment]
@eaplatanios could you put a build on this?

**Comment by anigasan** [comment]
@eaplatanios Were you able to put a build? Is there anything I need to fix?

**Comment by anigasan** [comment]
I have added necessary changes @eaplatanios, can you review it?

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 10]
```diff
@@ -496,6 +576,16 @@ final class MathOperatorTests: XCTestCase {
     }
 
     static var allTests = [
+        ("testAbs", testAbs),
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: It's generally a good idea to keep the function declarations above in the same order as this list. So, I would move your newly added `testAbs`-`testIdentity` after `testElementaryFunctions`.

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift [hunk 11]
```diff
@@ -58,6 +58,78 @@ final class MathOperatorTests: XCTestCase {
...
+        XCTAssertEqual(z, approxZ)
+    }
+    
+    func testZeros() {
+        let x = Tensor<Float>(zeros: [1])
+        let x1 = Tensor<Float>([0.0])
+        XCTAssertEqual(x, x1)
+    }
+    
+    func testLogSoftmax() {
```
Comment (eaplatanios) [inline Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Given that softmax should only be performed along the last axis, do you want to also add a higher-dimensional test case?

**Comment by eaplatanios** [comment]
Thanks @anigasan ! I made a couple of comments. Sorry for the delay but my schedule during the holidays is very unpredictable.

**Comment by anigasan** [comment]
It's perfectly alright! Thanks for reviewing :)

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift (hunk 10 above)
Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Will do!

File: Tests/TensorFlowTests/OperatorTests/MathTests.swift (hunk 11 above)
Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Will do!

Comment (anigasan) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: For the softmax implentation, do you mean higher degree tensor input?

Comment (eaplatanios) [reply Tests/TensorFlowTests/OperatorTests/MathTests.swift]: Yes, exactly. :)

**Comment by anigasan** [comment]
Hello, I have made the changes and added higher dimensional test case as well as reordered the list. Thanks for your assistance :)

**Comment by Shashi456** [comment]
@anigasan A higher dimensional tensor would be a 2-d tensor i.e matrix or a 3-d tensor, the input to the softmax test is still a 1-d tensor i.e a vector.

**Comment by anigasan** [comment]
Sorry about that, will make the fix

**Comment by anigasan** [comment]
@eaplatanios When you get the chance, can you please review the changes I have done?

**Comment by anigasan** [comment]
@eaplatanios @marcrasi if anyone of you get the time to review the changes I did, can you please do so 
as soon as possible. Sorry for pesstering you about this

**Comment by anigasan** [comment]
I have made the changes @marcrasi @eaplatanios, at the soonest convenience for you, can you please review the changes I have done. Thanks!

**Comment by anigasan** [comment]
@eaplatanios can you look at the changes I have done and recommend any feedback!
Thanks for your assistance and hope to keep close correspondence in the future

**Comment by anigasan** [comment]
Thank you so much for your assistance!
